// mongodb-core/lib/auth/mongo_credentials.js
'use strict';
function getDefaultAuthMechanism(ismaster) {
  if (ismaster) {
    if (Array.isArray(ismaster.saslSupportedMechs)) {
      return ismaster.saslSupportedMechs.indexOf('SCRAM-SHA-256') >= 0
        ? 'scram-sha-256'
        : 'scram-sha-1';
    }
    if (ismaster.maxWireVersion >= 3) {
      return 'scram-sha-1';
    }
  }
  return 'mongocr';
}
class MongoCredentials {
  constructor(options) {
    options = options || {};
    this.username = options.username;
    this.password = options.password;
    this.source = options.source || options.db;
    this.mechanism = options.mechanism || 'default';
    this.mechanismProperties = options.mechanismProperties;
  }
  equals(other) {
    return (
      this.mechanism === other.mechanism &&
      this.username === other.username &&
      this.password === other.password &&
      this.source === other.source
    );
  }
  resolveAuthMechanism(ismaster) {
    if (this.mechanism.toLowerCase() === 'default') {
      this.mechanism = getDefaultAuthMechanism(ismaster);
    }
  }
}
module.exports = { MongoCredentials };
